### PR TITLE
feat: add lint for recipe name

### PR DESCRIPTION
**Pull request by nichmor**
Checklist
* [x] Added a ``news`` entry
* [x] Regenerated schema JSON if schema altered (`python conda_smithy/schema.py`)

### CODE REVIEW

**Review by wolfv — APPROVED** [review]

**Comment by wolfv** [comment]
looks good. @nichmor maybe we should also send a PR that adds mypy to pre-commit?

**Comment by nichmor** [comment]
I like the idea!

File: conda_smithy/linter/rattler_linter.py [hunk 1]
```diff
@@ -101,3 +102,23 @@ def hint_noarch_usage(
 
         if no_arch_possible:
             hints.append(HINT_NO_ARCH)
+
+
+def lint_package_name(
```
[the remark this answers is not in the document]
Comment (wolfv) [reply conda_smithy/linter/rattler_linter.py:107]: That is correct. I think `rattler-build` also never accepts a package with an invalid name so it should basically never build successfully (which makes the lint a little less important for the rattler-build case).

[the remark this answers is not in the document]
Comment (nichmor) [reply conda_smithy/linter/rattler_linter.py:107]: can we skip in this case linting of the package name?

[the remark this answers is not in the document]
Comment (beckermr) [reply conda_smithy/linter/rattler_linter.py:107]: We only need to check the package section, not the context one.

[the remark this answers is not in the document]
Comment (wolfv) [reply conda_smithy/linter/rattler_linter.py:107]: Yes, but to properly lint the name we need to evaluate the jina, right?

We also need to double check that this works for multi-output recipes.

[the remark this answers is not in the document]
Comment (wolfv) [reply conda_smithy/linter/rattler_linter.py:107]: But more generally, rattler-build wouldn't allow you to build a package with an illegal name, so this lint might not be necessary. Do you still want it. @beckermr?

[the remark this answers is not in the document]
Comment (beckermr) [reply conda_smithy/linter/rattler_linter.py:107]: For v1 recipes we render the jinja2 before linting. We need to do the same for v2 recipes as well.

**Comment by nichmor** [comment]
@beckermr  added test that validates linting of recipe name using context value

File: tests/test_lint_recipe.py [hunk 2]
```diff
@@ -1396,6 +1396,38 @@ def test_recipe_name(self):
...
+        lints, _ = linter.lintify_meta_yaml(meta, is_rattler_build=True)
+        expected_message = (
+            "Recipe name has invalid characters. only lowercase alpha, "
+            "numeric, underscores, hyphens and dots allowed"
+        )
+        self.assertIn(expected_message, lints)
+
+        meta_with_context = {
+            "context": {"name": "mp++"},
+            "package": {"name": "${{ name }}"},
```
Comment (beckermr) [inline tests/test_lint_recipe.py:1410]:
```suggestion
            "context": {"blah": "mp++"},
            "package": {"name": "${{ blah }}"},
```

**Review by beckermr — CHANGES_REQUESTED** [review]
One nit to make the tests more robust.

**Comment by beckermr** [comment]
I've add a news file we can add items to as we merge PRs.

**Pull request merged**